### PR TITLE
ebs br: provide fsr warmup to tikv data volumes

### LINKED ISSUE #47271: ebs br: provide the option to do FSR warmup to TiKV volumes
## Enhancement
FSR solution to warmup AWS EBS snapshot is a solution with consistent post-restore performance. It's an addition to the existing hybrid warmup solution. When requested, TiDB BR pod needs to enable FSR before create data volume.

### LINKED ISSUE #47273: ebs br: don't change mysql.user table
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Use EBS snanpshot to restore a cluster, which contains user with name prefix of "cloud_"

### 2. What did you expect to see? (Required)
No content change to mysql. user table

### 3. What did you see instead (Required)
Some users in mysql.user are removed after restore

### PR DESCRIPTION
**Pull request by BornChanger**
### What problem does this PR solve?

Issue Number:
close #47271
close #47273 

Problem Summary:

### What is changed and how it works?

This PR addresses 2 issues
1. EBS snapshot restore can do FSR based warmup to tikv data volumes. A new parameter of`use-fsr` is introduced to  `br` command. Set it to `true` to use FSR to warmup tikv data volumes.
2. Get rid of unnecessary removal of some users from mysql.user table at the end of restore

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)
Run real EBS restore job and check the BR pod log to verify.
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/aws/ebs.go [hunk 1]
```diff
@@ -281,6 +281,73 @@ func (e *EC2Session) DeleteSnapshots(snapIDMap map[string]string) {
...
+func fetchTargetSnapshots(meta *config.EBSBasedBRMeta, targetAZ string) ([]*string, []*string) {
+	var sourceSnapshotIDs, availableZones []*string
+	for i := range meta.TiKVComponent.Stores {
+		store := meta.TiKVComponent.Stores[i]
+		for j := range store.Volumes {
+			oldVol := store.Volumes[j]
+			// Handle data volume snapshots only
+			if strings.Compare(oldVol.Type, "storage.data-dir") == 0 {
+				sourceSnapshotIDs = append(sourceSnapshotIDs, &oldVol.SnapshotID)
+				availableZones = append(availableZones, &targetAZ)
```
Comment (WangLe1321) [inline br/pkg/aws/ebs.go:343]: Why add `targetAZ` multiple times? I think the `availableZones` should only contain one item.

File: br/pkg/task/restore_ebs_meta.go [hunk 2]
```diff
@@ -236,7 +236,22 @@ func (h *restoreEBSMetaHelper) restoreVolumes(progress glue.Progress) (map[strin
...
+
+	// Turn on FSR for TiKV data snapshots
+	if h.cfg.UseFSR {
+		err = ec2Session.EnableDataFSR(h.metaInfo, h.cfg.TargetAZ)
+		if err != nil {
+			return nil, 0, errors.Trace(err)
+		}
+
+		log.Info("FSR for data volume snapshots are enabled")
+	}
```
Comment (WangLe1321) [inline br/pkg/task/restore_ebs_meta.go:253]: We should wait all the fsr finished before restore volumes.

Comment (BornChanger) [reply br/pkg/task/restore_ebs_meta.go:256]: fixed

File: br/pkg/aws/ebs.go [hunk 3]
```diff
@@ -281,6 +281,116 @@ func (e *EC2Session) DeleteSnapshots(snapIDMap map[string]string) {
...
 }
 
+// EnableDataFSR enables FSR for data volume snapshots
+func (e *EC2Session) EnableDataFSR(meta *config.EBSBasedBRMeta, targetAZ string) ([]*string, error) {
+	sourceSnapshotIDs, availableZones := fetchTargetSnapshots(meta, targetAZ)
+
+	log.Info("Start enable FSR for", zap.Int("snapshot number", len(sourceSnapshotIDs)), zap.Any("Snapshots", sourceSnapshotIDs), zap.String("available zone", targetAZ))
+
+	resp, err := e.ec2.EnableFastSnapshotRestores(&ec2.EnableFastSnapshotRestoresInput{
+		AvailabilityZones: availableZones,
```
Comment (WangLe1321) [inline br/pkg/aws/ebs.go:291]: `AvailabilityZones` should be `targetAZ`

File: br/pkg/aws/ebs.go [hunk 4]
```diff
@@ -281,6 +281,116 @@ func (e *EC2Session) DeleteSnapshots(snapIDMap map[string]string) {
...
+}
+
+// DisableDataFSR disables FSR for data volume snapshots
+func (e *EC2Session) DisableDataFSR(meta *config.EBSBasedBRMeta, targetAZ string) error {
+	sourceSnapshotIDs, availableZones := fetchTargetSnapshots(meta, targetAZ)
+
+	log.Info("Start disable FSR", zap.Int("snapshot number", len(sourceSnapshotIDs)), zap.Any("Snapshots", sourceSnapshotIDs), zap.String("available zone", targetAZ))
+
+	resp, err := e.ec2.DisableFastSnapshotRestores(&ec2.DisableFastSnapshotRestoresInput{
+		AvailabilityZones: availableZones,
```
Comment (WangLe1321) [inline br/pkg/aws/ebs.go:359]: same with above

File: br/pkg/aws/ebs.go [hunk 5]
```diff
@@ -281,6 +281,116 @@ func (e *EC2Session) DeleteSnapshots(snapIDMap map[string]string) {
...
+			log.Info("all snapshots fsr enablement is finished.")
+			return nil
+		}
+
+		// check pending snapshots every 1 minute
+		time.Sleep(1 * time.Minute)
+		log.Info("check snapshots not fsr enabled", zap.Int("count", len(pendingSnapshots)))
+		input := &ec2.DescribeFastSnapshotRestoresInput{
+			Filters: []*ec2.Filter{
+				{
```
Comment (WangLe1321) [inline br/pkg/aws/ebs.go:327]: we should add the `snapshot-id` filter. We should only watch the fsr we created.

**Review by WangLe1321 — APPROVED** [review]

File: br/pkg/task/restore_ebs_meta.go [hunk 6]
```diff
@@ -236,7 +238,21 @@ func (h *restoreEBSMetaHelper) restoreVolumes(progress glue.Progress) (map[strin
 			log.Error("failed to create all volumes, cleaning up created volume")
 			ec2Session.DeleteVolumes(volumeIDMap)
 		}
+
+		if h.cfg.UseFSR {
+			err = ec2Session.DisableDataFSR(snapshotsIDsMap)
```
Comment (3pointer) [inline br/pkg/task/restore_ebs_meta.go:243]: maybe record error

File: br/pkg/aws/ebs.go [hunk 7]
```diff
@@ -281,6 +281,155 @@ func (e *EC2Session) DeleteSnapshots(snapIDMap map[string]string) {
...
+func (e *EC2Session) EnableDataFSR(meta *config.EBSBasedBRMeta, targetAZ string) (map[string][]*string, error) {
+	snapshotsIDsMap := fetchTargetSnapshots(meta, targetAZ)
+
+	if len(snapshotsIDsMap) == 0 {
+		return snapshotsIDsMap, errors.Errorf("empty backup meta")
+	}
+
+	eg, _ := errgroup.WithContext(context.Background())
+
+	workerPool := utils.NewWorkerPool(e.concurrency, "enable snapshot FSR")
```
Comment (YuJuncen) [inline br/pkg/aws/ebs.go:294]: I think we don't need to restrict the concurrency here. Perhaps we can just enable all snapshots' FSR and polling the result at a central goroutine by `describe-fast-snapshot-restores` without filters.

File: br/pkg/task/restore_ebs_meta.go [hunk 8]
```diff
@@ -236,7 +238,21 @@ func (h *restoreEBSMetaHelper) restoreVolumes(progress glue.Progress) (map[strin
...
+		}
 	}()
+
+	// Turn on FSR for TiKV data snapshots
+	if h.cfg.UseFSR {
+		snapshotsIDsMap, err = ec2Session.EnableDataFSR(h.metaInfo, h.cfg.TargetAZ)
+		if err != nil {
+			return nil, 0, errors.Trace(err)
+		}
+
```
Comment (YuJuncen) [inline br/pkg/task/restore_ebs_meta.go:253]:
```suggestion 
```

File: br/pkg/aws/ebs.go (hunk 7 above)
Comment (YuJuncen) [reply br/pkg/aws/ebs.go:294]: ...If we enable the concurrency, and there are more disks than our concurrency limit, we may cost 2x time for enabling FSR.

**Review by YuJuncen — COMMENTED** [review]
rest lgtm

File: br/pkg/aws/ebs.go [hunk 9]
```diff
@@ -281,6 +281,155 @@ func (e *EC2Session) DeleteSnapshots(snapIDMap map[string]string) {
...
+
+	for availableZone := range snapshotsIDsMap {
+		workerPool.ApplyOnErrorGroup(eg, func() error {
+			log.Info("enable fsr for snapshots", zap.String("available zone", availableZone))
+			resp, err := e.ec2.EnableFastSnapshotRestores(&ec2.EnableFastSnapshotRestoresInput{
+				AvailabilityZones: []*string{&availableZone},
+				SourceSnapshotIds: snapshotsIDsMap[availableZone],
```
Comment (3pointer) [inline br/pkg/aws/ebs.go:301]: it seems race here. `availableZone` is a loop variable which cannot be capture by another goroutine. need a copy of `availableZone` here.

File: br/pkg/aws/ebs.go [hunk 10]
```diff
@@ -281,6 +281,155 @@ func (e *EC2Session) DeleteSnapshots(snapIDMap map[string]string) {
...
+
+	eg, _ := errgroup.WithContext(context.Background())
+
+	workerPool := utils.NewWorkerPool(e.concurrency, "disable snapshot FSR")
+
+	for availableZone := range snapshotsIDsMap {
+		workerPool.ApplyOnErrorGroup(eg, func() error {
+			resp, err := e.ec2.DisableFastSnapshotRestores(&ec2.DisableFastSnapshotRestoresInput{
+				AvailabilityZones: []*string{&availableZone},
+				SourceSnapshotIds: snapshotsIDsMap[availableZone],
```
Comment (3pointer) [inline br/pkg/aws/ebs.go:387]: ditto

File: br/pkg/aws/ebs.go (hunk 7 above)
Comment (BornChanger) [reply br/pkg/aws/ebs.go]: Enabling/Disabling FSR routines are executed in group of azs. The concurrency is changed to equal to the number of unique azs. So, in reality, the concurrency to enable/disable fsr is 3 or so.

File: br/pkg/aws/ebs.go [hunk 11]
```diff
@@ -281,6 +281,156 @@ func (e *EC2Session) DeleteSnapshots(snapIDMap map[string]string) {
...
+	}
+
+	eg, _ := errgroup.WithContext(context.Background())
+
+	workerPool := utils.NewWorkerPool(uint(len(snapshotsIDsMap)), "enable snapshot FSR")
+
+	for availableZone := range snapshotsIDsMap {
+		targetAZ := availableZone
+		workerPool.ApplyOnErrorGroup(eg, func() error {
+			log.Info("enable fsr for snapshots", zap.String("available zone", availableZone))
```
Comment (3pointer) [inline br/pkg/aws/ebs.go:299]:
```suggestion
			log.Info("enable fsr for snapshots", zap.String("available zone", targetAZ))
```

**Review by 3pointer — APPROVED** [review]
rest LGTM

File: br/pkg/aws/ebs.go (hunk 7 above)
Comment (YuJuncen) [reply br/pkg/aws/ebs.go]: If the pool size is always equal to the task, perhaps we can directly use `eg.Go( task )`?

**Review by YuJuncen — APPROVED** [review]

**Pull request merged**